traffic_ops/traffic_ops_golang/deliveryservice/acme.go

```diff
@@ -0,0 +1,363 @@
+		return
+	}
+	xmlID := inf.Params["xmlid"]
+
+	if userErr, sysErr, errCode := tenant.Check(inf.User, xmlID, inf.Tx.Tx); userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+
+	ctx, _ := context.WithTimeout(r.Context(), LetsEncryptTimeout)
+
+	err := renewAcmeCerts(inf.Config, xmlID, ctx, inf.User)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, err)
+	}
+
+	api.WriteRespAlert(w, r, tc.SuccessLevel, "Certificate for "+xmlID+" successfully renewed.")
+
+}
+
+func renewAcmeCerts(cfg *config.Config, dsName string, ctx context.Context, currentUser *auth.CurrentUser) error {
+	db, err := api.GetDB(ctx)
+	if err != nil {
+		log.Errorf(dsName+": Error getting db: %s", err.Error())
+		return err
+	}
+
+	tx, err := db.Begin()
+	if err != nil {
+		log.Errorf(dsName+": Error getting tx: %s", err.Error())
+		return err
+	}
+
+	userTx, err := db.Begin()
+	if err != nil {
+		log.Errorf(dsName+": Error getting userTx: %s", err.Error())
+		return err
+	}
+	defer userTx.Commit()
+
+	logTx, err := db.Begin()
+	if err != nil {
+		log.Errorf(dsName+": Error getting logTx: %s", err.Error())
+		return err
+	}
+	defer logTx.Commit()
+
+	dsID, certVersion, err := getDSIdAndVersionFromName(db, dsName)
+	if err != nil {
+		return errors.New("querying DS info: " + err.Error())
+	}
+	if dsID == nil || *dsID == 0 {
+		return errors.New("DS id for " + dsName + " was nil or 0")
+	}
+	if certVersion == nil || *certVersion == 0 {
+		return errors.New("certificate for " + dsName + " could not be renewed because version was nil or 0")
+	}
+
+	keyObj, ok, err := riaksvc.GetDeliveryServiceSSLKeysObjV15(dsName, strconv.Itoa(int(*certVersion)), tx, cfg.RiakAuthOptions, cfg.RiakPort)
+	if err != nil {
+		return errors.New("getting ssl keys for xmlId: " + dsName + " and version: " + strconv.Itoa(int(*certVersion)) + " :" + err.Error())
+	}
+	if !ok {
+		return errors.New("no object found for the specified key with xmlId: " + dsName + " and version: " + strconv.Itoa(int(*certVersion)))
+	}
+
+	err = base64DecodeCertificate(&keyObj.Certificate)
+	if err != nil {
+		return errors.New("decoding cert for XMLID " + dsName + " : " + err.Error())
+	}
+
+	acmeAccount := getAcmeAccountConfig(cfg, keyObj.AuthType)
+	if acmeAccount == nil {
+		return errors.New("No acme account information in cdn.conf for " + keyObj.AuthType)
+	}
+
+	client, err := GetAcmeClient(acmeAccount, userTx, db)
+	if err != nil {
+		log.Errorf(dsName+": Error getting acme client: %s", err.Error())
+		api.CreateChangeLogRawTx(api.ApiChange, "DS: "+dsName+", ID: "+strconv.Itoa(*dsID)+", ACTION: FAILED to add SSL keys with "+acmeAccount.AcmeProvider, currentUser, logTx)
+		return errors.New("getting acme client: " + err.Error())
+	}
+
+	renewRequest := certificate.Resource{
+		Certificate: []byte(keyObj.Certificate.Crt),
+	}
+
+	cert, err := client.Certificate.Renew(renewRequest, true, false)
+	if err != nil {
```

Comment:
Could we add a `nil` check for `cert` here?

Reply:
i dont think cert will ever be nil without an error but i threw it in